### DISCUSSION TITLE
Model Viewer Animation Question

### DISCUSSION THREAD

**Discussion by emincanyuksel** [google/model-viewer#4522] [Q&A] [answered]
Hello, I have a question. As you can see in the picture, there are 2 separate characters and actually both models have separate animations, but only 1 of them works, is there a way to run both of them?

![image](https://github.com/google/model-viewer/assets/125218696/14e0090d-0ead-492d-baef-c3d3676d41bf)

**Comment by zpennachi** [comment, marked as answer]
if you want both animations to play/loop simultaneously then yes! Entirely possible. You just want to make sure when exporting from blender you uncheck the "group by NLA track" in the animation dropdown under the GLB export settings.

If you want independent control over the animations (one plays when x happens, other plays when y happens) then you want to actually IGNORE what I said above, and then use the modelviewer API + some javascript to achieve said functionality. 

[Example:](https://www.thoughtful.fun/spooky) Based on the pose it looks like you're using Mixamo library. Here's an example of a single model that has multiple animations that can be triggered via hovering over the dots on the bottom of the page

**Reply by emincanyuksel** [reply]
Thank you very much, I tried what you said and it was solved.

**Reply by zpennachi** [reply]
Awesome! Glad I could help. I'm slowly assembling code snippets and video walkthrough tutorials going over these more nuanced/"advanced" bits of modelviewer/integrating pieces outside of modelviewer like Blender, Gestaltor, Mixamo, etc etc. If you have any other questions i'm more than happy to help.

**Reply by emincanyuksel** [reply]
Is it possible to apply the logic I wrote below?

**Comment by emincanyuksel** [comment]
Well, I have another question, can someone who knows answer it? Is it possible to do the FPS camera thing, that is, walk around the model and examine it?

**Reply by zpennachi** [reply]
I think you're describing AR mode. Again, very easy w/ modelviewer! If you want to test it out in the editor, just navigate to the first tab, open mobile view dropdown, click deploy mobile, and scan w/ your phone. If you want to include it in your scene, just leave the default copy/paste code the editor generates to include the button in your final scene.

**Reply by zpennachi** [reply]
https://www.zpennachi.com/ example: click the "view in your space" button on the 3d model to activate AR mode.

**Reply by emincanyuksel** [reply]
No, that's not what I mean. Is it possible to look at the model in first person using the WASD keys instead of navigating with the mouse?

**Reply by zpennachi** [reply]
Ah I see, so more like a traditional video game. I mean, in theory, yes a certain version of this is possible because modelviewer exposes a ton of [camera controls to the API,](
https://modelviewer.dev/docs/#entrydocs-stagingandcameras-attributes-cameraControls) but you'd have to build this functionality out. I think if you wanted to take this functionality any further, you may want to consider putting together your own three.js project/deployment.

**Reply by emincanyuksel** [reply]
In logic, yes, but if it were just as I said, it would be enough. I wish there was an example, I could work accordingly.

**Reply by zpennachi** [reply]
Modelviewer was made as a low barrier to entry option for web developers to quickly/easily incorporate 3d models in their projects. I think it's perfect for folks who have a solid understanding of javascript to build on top of, but if you do not have javascript skills, it can be intimidating.

This is why i'm putting together a code snippet library. I will include WASD camera controls on my list of functionalities to explore.

Also, i'm not affiliated w/ modelviewer in any way, just a fan and artist :)

If you want to speak to the man, myth, the legend themselves, @elalish is around and super helpful! <3

**Reply by emincanyuksel** [reply]
If this happens, it will be very good for me anyway. I hope he will help me like you did.

**Reply by zpennachi** [reply]
Just so you know @emincanyuksel , everything you've asked for so far is javascript IN ADDITION to modelviewer.dev. This is typically done by developers that you would hire to do the paid work. I don't think @elalish is going to write free code for you :) but they are extremely well-versed in modelviewer.dev, its limitations, and are great at offering guidance. 

I normally don't do this, but I put this together for you: https://codepen.io/zpennachi/pen/zYeGOVZ 

If you need any more help, I'm always available for hire :)

**Reply by emincanyuksel** [reply]
Thank you very much for doing this, actually this is the logic I want, so let me say this, the logic I want is simply to walk like a human, maybe there will be more with this code, I don't know JS, I wish I knew and could do it

**Reply by zpennachi** [reply]
I think if you don't know JS, but want this functionality, you should check out [spline.design](https://spline.design/)

The performance leaves a lot to be desired, and personally I do not use it, but it has starter templates that include wasd keys.

**Reply by emincanyuksel** [reply]
I will look here, I can show you exactly what I want to do, I think you can see the project I made in this **[link](https://kentkart.emincanyuksel.com.tr/kentkart_brt/)**. Actually, I want to walk instead of looking around with the mouse, so I was looking for help.

**Reply by zpennachi** [reply]
I understand your request, its beyond the scope of just offering advice, good luck friend.

**Reply by eyaler** [reply]
@emincanyuksel you may want to check out something like: https://gkjohnson.github.io/three-mesh-bvh/example/bundle/characterMovement.html
